### PR TITLE
add truncated digest length to encoding and decoding

### PR DESCRIPTION
**Pull request by candeira**
- jbenet's rationale: https://github.com/jbenet/multihash/issues/1#issuecomment-9178361
- I use this in my own https://github.com/candeira:ex_multihashing, which is a high level library for multihash generation and verification.

### CODE REVIEW

**Comment by zabirauf** [comment]
Thanks a lot for the changes Javier. I have started looking at the change and will pull them in. Will let you know if I have some comments on it.

File: lib/multihash.ex [hunk 1]
```diff
@@ -8,7 +8,7 @@ defmodule Multihash do
   require Monad.Error
   import Monad.Error
 
-  @type t :: %Multihash{name: String.t, code: integer, length: integer, digest: integer}
+  @type t :: %Multihash{name: atom, code: integer, length: integer, digest: integer}
   defstruct name: "", code: 0, length: 0, digest: 0
```
Comment (zabirauf) [inline lib/multihash.ex]: `name` should default to `:nil`

File: lib/multihash.ex [hunk 2]
```diff
@@ -41,7 +43,8 @@ defmodule Multihash do
   # Error strings
   @error_invalid_digest_hash "Invalid digest or hash"
   @error_invalid_multihash "Invalid multihash"
-  @error_invalid_length "Invalid length of provided hash function"
+  @error_invalid_length "Invalid length"
+  @error_invalid_trunc_length "Invalid truncation length is longer than digest"
```
Comment (zabirauf) [inline lib/multihash.ex]: The error doesn't sound right. I think it should be
`Invalid truncation length` as it is being returned if `l <=0 or l >= default_length`

File: lib/multihash.ex [hunk 3]
```diff
@@ -173,19 +191,23 @@ defmodule Multihash do
...
 
   @doc """
   Decode the multihash to %Multihash{name, code, length, digest} structure
   """
-  defp decode_internal([code: code, length: length], <<digest::binary>>) do
+  defp decode_internal([code: code, length: _default_length], <<digest::binary>>, length) do
     {:ok, name} = get_hash_function <<code>>
     Monad.Error.return %Multihash{
-      name: to_string(name) |> String.replace("_", "-"),
+      name: to_string(name) |> String.replace("_", "-") |> String.to_atom,
```
Comment (zabirauf) [inline lib/multihash.ex]: This should just be `name: name` as we are now directly using atom as name.

**Comment by zabirauf** [comment]
@candeira Change looks good, just added few minor comments.

File: lib/multihash.ex (hunk 3 above)
Comment (candeira) [inline lib/multihash.ex]: Right, thanks.

File: lib/multihash.ex (hunk 2 above)
Comment (candeira) [inline lib/multihash.ex]: Yep.

**Comment by candeira** [comment]
I've applied the suggested changes and rebased.Cheers!

**Comment by zabirauf** [comment]
Thanks a lot @candeira for the changes.

**Pull request merged**